edgexfoundry/go-mod-secrets
# feat(security): implement runtime token provider GetRawToken client API call for obtaining secret store token

Closes: #140

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>







If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-secrets/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x ] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [ ] I am not introducing a new dependency (add notes below if you are)  go/spiffe v2.0.0, same as the one in `edgex-go` in the server side
- [x ] I have added unit tests for the new feature or bug fix (if not, why?)
  <link to docs PR>

## Testing Instructions

- Run the edgex-compose's with edgex-go's spiffe token provider related services
- Use the following go code to test and if successful, you should obtain a secret store token:
```go
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/edgexfoundry/go-mod-core-contracts/v2/clients/logger"
	"github.com/edgexfoundry/go-mod-secrets/v2/pkg/token/runtimetokenprovider"
	"github.com/edgexfoundry/go-mod-secrets/v2/pkg/types"
)

func main() {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	socketPath := "/tmp/edgex/secrets/spiffe/public/api.sock"
	os.Setenv("SPIFFE_TRUST_DOMAIN", "edgexfoundry.org")
	testServiceKey := "security-test-client"

	runtimeTokenProvider := runtimetokenprovider.NewRuntimeTokenProvider(ctx, logger.NewMockClient(),
		types.RuntimeTokenProviderInfo{
			Enabled:         true,
			Protocol:        "https",
			Host:            "edgex-security-spiffe-token-provider",
			Port:            59841,
			SocketPath:      socketPath,
			RequiredSecrets: "redisdb",
		})

	token, err := runtimeTokenProvider.GetRawToken(testServiceKey)
	if err != nil {
		fmt.Println("failed to get token from spiffe-token provider: ", err.Error())
		os.Exit(1)
	}

	fmt.Println("OK: raw token = ", token)
}

```
## New Dependency Instructions (If applicable)
